### ISSUE TITLE
NodeFeature resource is namespaced - undefined behaviour if same node in multiple namespaces?

### ISSUE THREAD

**Issue by nobbs** [kubernetes-sigs/node-feature-discovery#1601]
Hi there👋🏻 

Stumbling into the following issue is probably going to happen to very few people, but nonetheless it might be good to report it. So, I'll try to describe the situation that led to it and what happened.

**What happened**:

In our case, migrating nfd from one namespace to another left outdated `NodeFeature` resources in the origin namespace which blocked node label updates even though the `NodeFeature` resources in the new namespace received updates.

**What you expected to happen**:

The `NodeFeature` resources are namespaced - the `nfd-api-controller` uses a sharedInformer, cf. https://github.com/kubernetes-sigs/node-feature-discovery/blob/2914bff8b38e841df90f5240a7f93323ff8e93f2/pkg/nfd-master/nfd-api-controller.go#L62-L66
that watches all namespaces for changes to the nfd resources. As far as I can tell from looking into the code, there is no check which namespace those resources are living in. If the `NodeFeature` resources have to be namespaced (not sure why though), maybe some kind of priority for the namespace nfd is running in should be considered? 

**How to reproduce it (as minimally and precisely as possible)**:

1. We initially deployed nfd as a requirement for nvidias gpu-operator months ago, in the same namespace, called `gpu-operator`. We kept it up to date, switched later to the new API with the `NodeFeature` CRD
2. At some point, we decided to migrate the nfd to its own namespace, `node-feature-discovery`. This migration did not remove the previously created `NodeFeature` resources in the `gpu-operator` namespace, but these were no longer updated. New `NodeFeature` resources were created in the `node-feature-discovery` namespace and kept up to date
3. Weeks later, we noticed that the Node labels are not being updated (kernel version was the giveaway). Strangely, the `NodeFeature` resources in `node-feature-discovery` were up to date and had the labels we would've expected on the nodes, the nodes on the other hand were not updated.
5. By chance, I've spotted the old `NodeFeature` resources in the `gpu-operator` namespace that matched the observed node labels. Deleting these resources, the node labels started to get properly updated again.

I'm not sure why exactly it preferred the objects from one namespace over the other - that was possibly just a race condition.

**Comment by marquiz** [comment]
Thanks @nobbs for the report. This is a recognized issue. There are measures we've taken to mitigate the problem of stale objects and labels. Which version of NFD were you using?

1. #1491 (released in NFD v0.15) added owner references to NodeFeature objects so that stale objects are GC'd (by Kubernetes) when the nfd worker pod that created it is deleted.
2. #1532 (will be released in NFD v0.16) is somewhat related (albeit does not help with stale NF objects) added a post-delete hook in the Helm chart to remove all NFD-generated labels when NFD gets uninstalled.
3. #1592 is an upcoming feature to limit/select the k8s namespaces where to watch NodeFeature objects.

For kustomize-based deployments doing `kubectl delete -k ...` deletes the nfd namespace and the stale NodeFeature objects as well. It will not cleanup the labels, though.

**Comment by nobbs** [comment]
Hey @marquiz - thanks for the info.

As we tried to track the releases nfd quite closely - most of the time delayed by 2-4 weeks - and the timeline the issue started, I'm pretty sure we migrated from one namespace to another either on one of the last v0.13.x patch releases or even on v0.14.0. The obsolete NodeFeature resources definitely had no owner references set at the time I've deleted them.

Based on the PRs you've linked it looks like the chance of running into this issue is basically naught with the recent / upcoming versions. Free to close this issue as resolved. :)
